### PR TITLE
V6

### PR DESCRIPTION
**Pull request by bluebill1049**
### 📖 Documentation

- [x] Update V6 documenation: https://github.com/react-hook-form/react-hook-form-website/pull/252
- [x] New Page TS 
  - [x] scope/limit exposed Types
- [x] Include selection under API page to change version
- [x] Migration Guide V5 -> V6

### 🚨 Breaking changes

- [x] remove `validationSchema` and embrace validation resolver
https://github.com/react-hook-form/react-hook-form-resolvers

**Reason:** we would love to support all kinds of schema validation instead of just Yup, also it's not fair to carry that extra logic for users who don't use schema at all. also perhaps there are lighter options than only `Yup` as well.

| Name | Size | Url |
|--|--|--|
| Superstruct | 4.5KB | [link](https://github.com/ianstormtaylor/superstruct) |
| Yup | 20KB | [link](https://github.com/jquense/yup) |
| Joi | 29KB | [link](https://github.com/hapijs/joi) |

```diff
- useForm({ validationSchema: schema })
+ useForm({ resolver: yupResolver(schema) })
```

```ts
import React from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from 'react-hook-form-resolvers';

const SignupSchema = yup.object().shape({
  name: yup.string().required(),
  age: yup.number().required(),
});

const App = () => {
  const { register, handleSubmit } = useForm({
    resolver: yupResolver(SignupSchema), // yup, joi and even your own.
  });

  return (
    <>
      <form onSubmit={handleSubmit(d => console.log(d))}>
        <label>Test</label>
        <input name="test" ref={register} />
        <input type="submit" />
      </form>
    </>
  );
};
```

---------------------

- [x] reanme `validationResolver` to resolver
- [x] reanme `validationContext` to context
- [x] reanme `validateCriteriaMode` to `criteriaMode`

**Reason:** less for the user to type, consistent with the rest of the APIs. This is a generic validation library, it makes sense to keep naming generic and validation should already be the context for this whole lib.  

```diff
- validationResolver,
- validationContext,
+ resolver,
+ context,
```

---------------------

- [x] `Controller` props `onChange` change from array arguments to object arguments

**Reason:** This should be the expected behavior for function's `arguments`. 

```diff
- onChange = {[e] => console.log(e)};
+ onChange = {e => console.log(e)};
```

---------------------

- [x] `triggerValidation` rename to `trigger`

**Reason:** Less for a user to type and consistent with the rest of the APIs. However, if this having issue or conflict with your existing code base, you have the option to rename it in the deconstructor.

```diff
- triggerValidation();
+ trigger();
```

---------------------

- [x] `FormContext` rename to `FormProvider`

**Reason:** Align name with all those big libraries such as Redux, styled-components and etc.

```diff
- <FormContext />
+ <FormProvider />
```

- [x]  export `FormContext`

**Reason:** Some users may want to use `FormContext.Consumer`.

```jsx
import { FormContext } from 'react-hook-form';

// in your connected component
render() {
  return (
    <FormContext.Consumer>
      {({ register, handleSubmit }) => {
        // do something with the useForm methods here
      }}
    </FormContext.Consumer>
  )
};
```

---------------------

- [x] remove `nest` option for `watch`& `getValues`, so data return from both methods will be in `FormValues` shape.

**Reason:** Consistency for form values and better type support.

```diff
- getValues({ nest: true }); // { test: { data: 'test' }}
- watch({ nest: true });  // { test: { data: 'test' }}
+ getValues(); // { test: { data: 'test' }}
+ watch();  // { test: { data: 'test' }}
````

###  🥃 New Features

- [x] TS: `NestedValue` 

**Reason**: The field value can be array/object such as Material-UI Multiple Select and custom registered field (Custom Register).

Currently, errors object expects an array value to always be multiple fields, but it can also be a single field with an array value. To solve this, it supports NestedValue to determine if it is the field value (array/object).

```ts
import { useForm, NestedValue } from 'react-hook-form';

type FormValues = {
  key1: string;
  key2: number;
  key3: NestedValue<{
    key1: string;
    key2: number;
  }>;
  key4: NestedValue<string[]>
};

const { errors } = useForm<FormValues>();

errors?.key1?.message // no type error
errors?.key2?.message // no type error
errors?.key3?.message // no type error
errors?.key4?.message // no type error
```

---------------------

- [x] `useWatch` (new) subscribe to registered inputs.

**Reason:** This could isolate the re-render within a particular component without trigger re-render at the form level. This is different from `watch` API, which will re-render at the root of `useForm`

```ts
<input name="test" ref={register} />

function IsolateReRender() {
  const { state } = useWatch({
    name: 'test',
    control,
    defaultValue: 'default'
  });

  return <div>{state}</div>
}
```

---------------------

- [x] `getValues()` support array of field names

 **Reason:** API Consistency with `watch`

```ts
+ getValues(['test', 'test1']); // { test: 'test', test1: 'test1' }
````

---------------------

- [x] `useForm({ mode: 'all' })` support all validation

 **Reason:** so each input can validate with blur and change. https://github.com/react-hook-form/react-hook-form/issues/1222

```ts
+ useForm({ mode: 'all' })
````

### CODE REVIEW

File: src/data/jp/migrateV5ToV6.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
+import CodeArea from "../../components/CodeArea"
+import validationSchema from "../../components/codeExamples/validationSchema"
+import * as React from "react"
+import typographyStyles from "../../styles/typography.module.css"
+
+export default (
```
Comment (bluebill1049) [inline src/data/jp/migrateV5ToV6.tsx:6]: @kotarella1110 can you please translate this content?

File: src/data/kr/migrateV5ToV6.tsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
+import CodeArea from "../../components/CodeArea"
+import validationSchema from "../../components/codeExamples/validationSchema"
+import * as React from "react"
+import typographyStyles from "../../styles/typography.module.css"
+
+export default (
+  <>
```
Comment (bluebill1049) [inline src/data/kr/migrateV5ToV6.tsx:7]: @adhrinae can you please help translate the following content? thanks

File: src/data/V5/es/api.tsx [hunk 3]
```diff
@@ -902,90 +902,36 @@ export default {
...
     ),
   },
-  ErrorMessage: {
-    title: "ErrorMessage",
-    description: (
-      <p>
-        Un componente simple para representar el mensaje de error del input
-        asociado.
-      </p>
+  table: {
```
Comment (bluebill1049) [inline src/data/V5/es/api.tsx:905]: @kotarella1110 That's the reason made the build failed, i am fixing this. 👍

File: src/data/kr/api.tsx [hunk 4]
```diff
@@ -612,14 +612,65 @@ export default {
...
           </td>
           <td>
-            <code className={typographyStyles.typeText}>
-              React.ElementType | string
-            </code>
+            <code className={typographyStyles.typeText}>React.ElementType</code>
           </td>
-          <td>✓</td>
+          <td></td>
+          <td>
```
Comment (bluebill1049) [inline src/data/kr/api.tsx:618]: @adhrinae please help with the following when you have free time. i used google translate...

File: src/data/jp/api.tsx [hunk 5]
```diff
@@ -627,31 +627,99 @@ export default {
...
-            <code className={typographyStyles.typeText}>
-              React.ElementType | string
-            </code>
+            <code className={typographyStyles.typeText}>Object</code>
           </td>
           <td>✓</td>
           <td>
-            制御されたコンポーネント。 例: <code>as="input"</code> または{" "}
-            <code>{`as={<TextInput />}`}</code>
+            <code>control</code> オブジェクトは <code>useForm</code>{" "}
```
Comment (bluebill1049) [inline src/data/jp/api.tsx:637]: @kotarella1110 please take a look when you got a chance for the translation.

File: src/data/kr/api.tsx (hunk 4 above)
Comment (adhrinae) [reply src/data/kr/api.tsx:618]: Sure 🙂 I'm thinking to get things done at once.

Comment (bluebill1049) [reply src/data/kr/api.tsx:618]: thank you! that would be great. i would let you know once everything is ready.

File: src/data/V5/es/api.tsx (hunk 3 above)
Comment (kotarella1110) [reply src/data/V5/es/api.tsx]: @bluebill1049 oh... sorry! thanks!

**Comment by bluebill1049** [comment]
@pmaier983 `npm run lint` hangs at my mac, so i have disabled it for now.

File: src/components/codeExamples/testingForm.ts [hunk 6]
```diff
@@ -15,39 +15,32 @@ export default function App({ login }) {
...
-              value: 5,
-              message: "min length is 5"
-            }
-          })}
-          type="password"
-          placeholder="password"
-        />
-        {errors.password && <span role="alert">{errors.password.message}</span>}
-      </section>
+      <label htmlFor="email">email</label>
```
Comment (bluebill1049) [inline src/components/codeExamples/testingForm.ts:18]: @keiya01 I did reduce couple lines with the test example to keep the page a bit shorter, let me know if you ok with that.

File: src/components/codeExamples/testingForm.ts [hunk 7]
```diff
@@ -15,39 +15,32 @@ export default function App({ login }) {
...
-              message: "Entered value does not match email format"
-            }
-          })}
-          type="email"
-          placeholder="example@mail.com"
-        />
-        {errors.email && <span role="alert">{errors.email.message}</span>}
-        <label htmlFor="password">password</label>
-        <input
-          id="password"
```
Comment (keiya01) [inline src/components/codeExamples/testingForm.ts]: id="password" is necessary because it is used in test.

Comment (bluebill1049) [reply src/components/codeExamples/testingForm.ts]: will add them back 🙏

Comment (bluebill1049) [reply src/components/codeExamples/testingForm.ts]: i thought you are only using the `name`?

Comment (bluebill1049) [reply src/components/codeExamples/testingForm.ts]: nvm, it's for the label 👍

Comment (keiya01) [reply src/components/codeExamples/testingForm.ts]: Yes, it's for the label :)

Comment (keiya01) [reply src/components/codeExamples/testingForm.ts]: Thank you.

Comment (bluebill1049) [reply src/components/codeExamples/testingForm.ts]: by the way, you don't have to change the codesandbox :) it's only for the webiste

Comment (keiya01) [reply src/components/codeExamples/testingForm.ts]: OK. Thank you :)

File: src/data/en/advanced.tsx [hunk 8]
```diff
@@ -539,18 +538,33 @@ export default {
...
 
-        <p>
-          Note that you have to wait for finishing submission by using{" "}
-          <code>waitFor</code> or <code>find*</code> method. Because{" "}
-          <code>handleSubmit</code> method is invoked with asynchronous. And you
-          are as possible as use <code>*ByRole</code> method when you want to
-          query element. Because users recognize role(like button, textbox) that
-          UI have it when users use your features.
-        </p>
+        <p>The following criteria are what we try to cover with the tests:</p>
```
Comment (bluebill1049) [inline src/data/en/advanced.tsx:544]: FYI @keiya01 I did some cleanup, I will translate to other languages now.

Comment (keiya01) [reply src/data/en/advanced.tsx:544]: OK! I will check it out later.

**Comment by bluebill1049** [comment]
@adhrinae could you please verify Korean translation?

**Comment by adhrinae** [comment]
@bluebill1049 Yeah, sure. Should I push the changes to this branch?

**Comment by bluebill1049** [comment]
yes please @adhrinae

File: src/data/ts.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import en from "./en/ts"
+import es from "./es/ts"
+import jp from "./jp/ts"
+import kr from "./kr/ts"
+import pt from "./pt/ts"
+import ru from "./ru/ts"
+import zh from "./zh/ts"
+
+export { en, es, jp, kr, pt, ru, zh }
```
Comment (adhrinae) [inline src/data/ts.ts:9]: @bluebill1049 I've tried to support `currentLanguage` prop in order to apply i18n on `tsPage`.

Comment (bluebill1049) [reply src/data/ts.ts:9]: awesome! thank you 🙏

File: package.json [hunk 10]
```diff
@@ -1,43 +1,51 @@
...
+  "version": "6.0.0",
   "author": "beier luo",
   "engines": {
     "node": "12.x"
   },
   "dependencies": {
-    "gatsby": "2.21.21",
-    "gatsby-image": "2.4.3",
-    "gatsby-plugin-google-analytics": "2.3.1",
+    "@hookform/devtools": "2.0.0-beta.1",
```
Comment (pmaier983) [inline package.json:10]: Is there any reason hookform/devtools is not in devDependencies. (Also we should bump to 1.2.1). Will create PR in the near future.

**Comment by adhrinae** [comment]
Korean translation for V6 is done.
I will revise V5 docs later after launching V6.

File: package.json (hunk 10 above)
Comment (bluebill1049) [reply package.json:10]: its a specail build, because normally devtool export null in prod.

**Comment by bluebill1049** [comment]
> Korean translation for V6 is done.
> I will revise V5 docs later after launching V6.

Thank you very much @adhrinae <3

File: src/components/codeExamples/testingForm.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 export const step1 = `module.exports = {
-  setupFiles: ["<rootDir>/setup.ts"]
+  setupFilesAfterEnv: ["<rootDir>/setup.ts"]
```
Comment (keiya01) [inline src/components/codeExamples/testingForm.ts:2]: @bluebill1049  I fixed code example.

Comment (bluebill1049) [reply src/components/codeExamples/testingForm.ts:2]: Thanks @keiya01 today is the day ❤️

Comment (keiya01) [reply src/components/codeExamples/testingForm.ts:2]: I’m so nervous🥳

**Pull request merged**